src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -338,6 +338,17 @@ where
                                     break;
                                 }
                                 Some(chunk) => {
+                                    if chunk.cardinality() == 0 {
+                                        tracing::warn!("barrier_snapshot_read chunk is empty");
+                                    }
+
+                                    let row = chunk.rows().next().unwrap();
```

Comment:
Hmm, this implementation actually reads 1 chunk from storage, and then take out the first row. Is it possible to read 1 row from storage?

Reply:
I think it's the simplest approach.

Here we are reading an already created stream, whose granularity in at the chunk level.

To read 1 row instead, it means we either have to:
1. Start a new snapshot read which reads at a row level granularity (we could lose the tombstone iteration process, defeating the purpose of this optimization).
2. The original snapshot read has to yield at row level granularity, then we have to manually batch and reconstruct the chunk on top of the iterator.

Both don't really seem ideal.

I think we can trade-off efficiency, and gain code-simplicity by using the approach in this PR.
After all the dominant factors are from tombstone and amplification, so it won't affect the backfill efficiency much.

Reply:
With the assumption that tombstones are mostly right ahead, I think the cost of reading one row and one chunk are almost the same. If we have already got a chunk, sending them as a whole downstream is fine. I think we don’t need to do this just for the sake of sticking to the literal meaning.

Reply:
If amplification is around `1000x`, then sending the full chunk could be `256,000` records per epoch, if chunk size 256.

Reply:
Any further comments?

Reply:
Another approach is to share an atomic boolean stop signal for the snapshot stream, where it's checked before appending every row. It's normally `false` but set to `true` here. 😄 Not a big deal because I believe what @chenzl25 said makes sense.

Reply:
With some discussion, sending 1 row to downstream LGTM, because in the ideal situation, if we could record tombstone in backfill state, we would send no data downstream. However, to skip tombstone, we choose to send 1 row instead which is closest to zero. Make sense.